### PR TITLE
Use ruby unicode normalize to avoid libidn C problems and heavy legacy ruby code

### PR DESCRIPTION
**Pull request by jarthod**
This is a fix for #408 and the beginning of the simplification and modernization of the IDNA code. It does:
- Replaces `libidn` unicode normalize implementation by ruby's, to avoid the null terminated string problem and support more recent unicode versions.
- Replaces `Pure` unicode normalize by ruby's, to avoid hundreds of line of slow and legacy ruby code and support more recent unicode versions.

The result in terms of performance is the following (benchmark code commited, but needs to be run on master to measure the legacy "pure" code):
```
#              user     system      total        real
# pure     1.325948   0.000000   1.325948 (  1.326054)
# libidn   0.058067   0.000000   0.058067 (  0.058069)
# ruby     0.325062   0.000000   0.325062 (  0.325115)
```

The native ruby implementation is 5-6 times **slower** than `libidn`, but also 5 times **faster** than the existing `Pure` one.

Considering this is just a small part and the rest of the ruby code is more than one order of magnitude slower (doing a full `normalize` in the `simple.rb` benchmark for example takes 8.6 seconds for the same 100k iterations on my machine), I think this is more than enough, no need to try to shave off microseconds in the `unicode_normalize` code if the rest is a lot of ruby.

And we benefit from the native ruby function which is going to be maintained and improved.
Specs have been added to cover the problematic `\x00` case.

### CODE REVIEW

File: lib/addressable/idna/pure.rb
Comment (jarthod) [inline lib/addressable/idna/pure.rb:82]: And we need to do the normalise while we have an unicode string so before the conversion to `ASCII_8BIT`
Also the normalize should be before the split according to http://www.unicode.org/reports/tr46/#ProcessingStepNormalize.
So in the end it's both simpler and more correct this way :)

File: lib/addressable/uri.rb
Comment (jarthod) [inline lib/addressable/uri.rb:56]: Unrelated minor fix to remove the following warning: `uri.rb:417: warning: character class has duplicated range: /[^a-zA-Z0-9\-\.\_\~\!\$\&\'\(\)\*\+\,\;\=\:\@\[\:\]]` because the column `:` is already in `PCHAR`

File: spec/addressable/idna_spec.rb
Comment (jarthod) [inline spec/addressable/idna_spec.rb:46]: The ruby normalize function requires a proper Unicode string which was not the case before, so even though this was not the case anywhere, I though I would add a spec to make sure it still works even if the user (or other code) passes a wrongly encoded string (`ASCII-8BIT` in this example thanks to `.b`)

File: spec/addressable/idna_spec.rb
Comment (jarthod) [inline spec/addressable/idna_spec.rb:262]: Fails with `Native` before the fix.

File: spec/addressable/idna_spec.rb
Comment (jarthod) [reply spec/addressable/idna_spec.rb:43]: This does not match any line in this file

File: lib/addressable/uri.rb
Comment (dentarg) [reply lib/addressable/uri.rb:56]: Isn't this already done in https://github.com/sporkmonger/addressable/pull/438?

Comment (jarthod) [reply lib/addressable/uri.rb:56]: Ah indeed, I must have forked from and outdated local copy, thanks for pointing this out I'll rebase :)

Comment (jarthod) [reply lib/addressable/uri.rb:56]: Actually I was up to date, the column is back and I'm not even sure why :thinking:. I checked the git commit log, I can see this commit which remove the column: dd84176c2ad6c71bbec4b66bfc694d80969149ae and then this commit: 178bcd97fc77b342a3bda6f5b6e17f8ea344bd51 added a `freeze` but somehow the column is back, even in the "previous" version in red...

`git log -p lib/addressable/uri.rb`:
image

I dont... :shrug: Any idea how this could happen? Anyway I think we can keep this change to put it back in the codebase then.

File: benchmark/unicode_normalize.rb
Comment (jarthod) [reply benchmark/unicode_normalize.rb:14]: Just a benchmark file and it's more readable this way IMO, but let me know if you really want to follow

File: lib/addressable/template.rb
Comment (jarthod) [inline lib/addressable/template.rb:912]: According to [the RFC](https://www.rfc-editor.org/rfc/rfc6570.txt), this normalization is actually supposed to be NFC form, not NFKC:
> If a value is provided by a user, such as via a data-entry dialog, then the string SHOULD be normalized as Normalization Form C (NFC: Canonical Decomposition, followed by Canonical Composition) prior to being used in expansions by a template processor.

It's easy to do now with ruby so instead of using the global function I'm switching to direct ruby call. Also I skip the normalize if the input string is not Unicode, this is faster and avoids potential exceptions. Finally I used a recursive call for the array and hash keys/values, this way the same behavior (NFC and only normalize if Unicode) is shared between all types.

File: lib/addressable/uri.rb
Comment (jarthod) [inline lib/addressable/uri.rb:484]: Just for consistency, we use this way in all other places.

File: lib/addressable/uri.rb
Comment (jarthod) [inline lib/addressable/uri.rb:690]: Same thing for the both calls in URI we should use NFC form instead of NFKC (only for IDNA), so I switched to the simple ruby call. `:nfc` is the default form but I let it as an argument to show that it's a deliberate choice.

File: spec/addressable/idna_spec.rb
Comment (jarthod) [inline spec/addressable/idna_spec.rb:256]: This function is no longer needed and there's no need to test ruby own implementation.

**Comment by jarthod** [comment]
Ok I've fixed some encoding issues with template on ruby 2.7- (passed with some ascii strings it was failing) and most hound warnings. I've left comments to explain the changes. All tests are passing. Let me know if you have any comment.

As another validation step, I'm gonna run this version in my product (https://updown.io) which converts 150k+ URl regularly to see if I spot anything.

File: lib/addressable/uri.rb
Comment (dentarg) [reply lib/addressable/uri.rb:56]: I #449 was opened the same day #438 was merged so #449 probably didn't have the latest changes and we did not notice in review 😞

Comment (dentarg) [reply lib/addressable/uri.rb:56]: Plus GitHub not understanding it was a conflict? Anyway, I asked @sporkmonger to double check repo settings in that area (as I don't have access to that myself)

File: lib/addressable/template.rb
Comment (dentarg) [reply lib/addressable/template.rb:912]: >I skip the normalize if the input string is not Unicode

Is it possible for the input string to be Unicode but not UTF-8? 🤔 (Maybe not a realistic scenario even if it can happen?)

File: benchmark/unicode_normalize.rb
Comment (dentarg) [reply benchmark/unicode_normalize.rb:14]: I agree with you and I would raise this limit if I could, but let's follow the bot so it doesn't complain in the future.

(Maybe `# rubocop:disable Metrics/LineLength` comments work? (don't forget to enable the cop again after the section where it is needed.)

File: spec/addressable/idna_spec.rb
Comment (dentarg) [reply spec/addressable/idna_spec.rb:43]: I don't follow, is the complaint from the bot not correct? :)

**Review by dentarg — APPROVED** [review]
Looks good to me, though I'm no expert at all in this area! I trust our tests and your testing @jarthod :-) Thank you for this very clear pull request.

File: spec/addressable/idna_spec.rb
Comment (jarthod) [reply spec/addressable/idna_spec.rb:43]: What I mean is that all the specs in this file (and most in other spec files) are written with this indentation, so they are all "incorrect" for @houndci-bot. So if I change it to folow this recommendation, it'll look unaligned with other tests and waste a lost of _precious_ chars toward the 80 chars limit. So if we try to make it consistent and fix the other specs, we'll have lines "too long" again and will need even more fixes. Overall making the code much harder to read so a bad idea to foolow it IMO.

File: lib/addressable/template.rb
Comment (dentarg) [reply lib/addressable/template.rb:912]: Ah, this is in `Addressable::Template`, maybe @therabidbanana want to chime in?

I'm really not familiar with templates but my feeling is that the change is okay.

File: benchmark/unicode_normalize.rb
Comment (dentarg) [reply benchmark/unicode_normalize.rb:14]: I don't know. Maybe don't please the bot and I will try my to get access to repo settings, to remove the bot. :)

**Comment by jarthod** [comment]
> Looks good to me, though I'm no expert at all in this area! I trust our tests and your testing @jarthod :-) Thank you for this very clear pull request.

Thanks! @brasic maybe you also have an opinion or feedback on this PR?

> I was thinking about a major version bump, but I guess this very much a bug fix? Unless there are some very strange edge cases we don't know about and people rely on.

I don't think this deserves a major bump, it's more of a bugfix indeed.
#491 and #247 definitely will though :)

File: spec/addressable/idna_spec.rb
Comment (dentarg) [reply spec/addressable/idna_spec.rb:43]: Thanks for explaining. Best course of action seems to ignore the bot, try to remove it, and if rules are introduced again, also update the code when introducing rules.

**Comment by dentarg** [comment]
> #491 and #247 definitely will though :)

Hehe, yes, was thinking that too :)

**Comment by dentarg** [comment]
> #491 and #247 definitely will though :)

Hehe, yes, was thinking that too :)

File: spec/addressable/idna_spec.rb
Comment (jarthod) [reply spec/addressable/idna_spec.rb:43]: I believe so too :) for the record I am not opposed to this kind of bot, but I am definitely against applying them "blindly" when it makes the code less readable and/or less consistent.

File: lib/addressable/uri.rb
Comment (jarthod) [reply lib/addressable/uri.rb:56]: Weirdest to me is that it's not only Github here, even in plain git (as in my git log screenshot) we can clearly see the "before" (red) line in https://github.com/sporkmonger/addressable/commit/178bcd97fc77b342a3bda6f5b6e17f8ea344bd51 being different than the "after" (green) line in https://github.com/sporkmonger/addressable/commit/dd84176c2ad6c71bbec4b66bfc694d80969149ae. And there's no other comits on this file in between. I have no idea how this can even be possible.

Anyway I suppose it's ok to keep this change here then.

File: benchmark/unicode_normalize.rb
Comment (jarthod) [reply benchmark/unicode_normalize.rb:14]: Ok let's ignore the bot for now, and if you want I can try to have a look at adding a `.hound.yml` in another PR (to keep it but tune it down). Not needed if you aim at removing it though.

File: lib/addressable/uri.rb
Comment (dentarg) [reply lib/addressable/uri.rb:56]: Yeah we'll keep it here 

The merge commit diff is correct: https://github.com/sporkmonger/addressable/commit/7ecf75146c1a53b089150ccc33c909214d728323

Comment (jarthod) [reply lib/addressable/uri.rb:56]: Interesting, I don't see this commit in the git log for `lib/addressable/uri.rb`: https://github.com/sporkmonger/addressable/commits/main/lib/addressable/uri.rb

And when I check it with:
```bash
> git show 7ecf75146c1a53b089150ccc33c909214d728323
commit 7ecf75146c1a53b089150ccc33c909214d728323
Merge: 068f673 41f12dd
Author: Bob Aman <sporkmonger@users.noreply.github.com>
Date:   Fri Jul 29 20:35:38 2022 -0700

    Merge pull request #449 from okeeblow/freeze_concatenated_strings
    
    Explicitly freeze concatenated-`String` constants to unbreak on non-main `Ractor`s.

>
```
It's "empty", there's no file changed. There's probably many things I still don't get about `git` :smiling_face_with_tear:

Comment (dentarg) [reply lib/addressable/uri.rb:56]: Yeah, I just [googled this](https://stackoverflow.com/questions/43138569/how-to-get-git-log-p-to-show-changes-in-merge-commits) and I learned something too here :) with `-m` (or the longer variant `--diff-merges=on`) `git log`/`git show` shows me the diff

```
git log -1 --diff-merges=on 7ecf751
git show --diff-merges=on 7ecf751
```

Comment (jarthod) [reply lib/addressable/uri.rb:56]: Oh interesting, thanks :)

File: lib/addressable/template.rb
Comment (sporkmonger) [reply lib/addressable/template.rb:912]: I'm inclined to agree that this should be changed. It looks like there may be a similar issue with RFC 5890?

> A "U-label" is an IDNA-valid string of Unicode characters, in Normalization Form C (NFC) and including at least one non-ASCII character, expressed in a standard Unicode Encoding Form (such as UTF-8).

A subtle change like this probably deserves some extra attention in tests.

File: lib/addressable/idna/pure.rb
Comment (sporkmonger) [inline lib/addressable/idna/pure.rb:69]: Should this still be NFKC?

Comment (jarthod) [reply lib/addressable/idna/pure.rb:69]: Good question and tricky answer. As far as I understand in [IDNA2008+UTC#46](https://unicode.org/reports/tr46/) there is the purely unicode normalization normalization step which is NFC, but there is also the IDNA mapping table, which is based on NFKC Casefold transformation. So in the end it does look similar to NFKC with some variations.

In IDNA2003 the "normalization" step was called "STRINGPREP" and was equivalent to NFKC (https://www.rfc-editor.org/rfc/rfc3491#ref-STRINGPREP):
> This profile specifies using Unicode normalization form KC, as described in [[STRINGPREP](https://www.rfc-editor.org/rfc/rfc3491#ref-STRINGPREP)].

Which is why [the libidn library only provides NFKC normalization functions](https://www.gnu.org/software/libidn/manual/libidn.html#Unicode-Normalization).

So in the end as we don't have the official mapping for the moment, NFKC is closer to the expected result (and was previously used), so I thought it would be better to keep it like that for the moment. This will probably change in #491 once we have the mapping table and all the conformance tests.

File: lib/addressable/template.rb
Comment (jarthod) [reply lib/addressable/template.rb:912]: > I'm inclined to agree that this should be changed

Do you mean changed back to accepting any encoding and treating it as UTF-8? or keeping the change that I made?

> It looks like there may be a similar issue with RFC 5890/5895?

I believe yes if a user passes in a UTC-16 encoded string for domain name, it'll probably raise. If we want to make this more strict in the IDNA part we could accept any input encoding and re-`encode` it as UTF-8 (instead of `force_encoding`). This is much more likely to break existing people code though if there are invalid or inconvertible chars. Probably something we can discuss more in #491

> A subtle change like this probably deserves some extra attention in tests.

Good point I'll add some more tests around this.

Comment (jarthod) [reply lib/addressable/template.rb:912]: Ok I've just added two specs covering the new case: non-unicode strings passed which should **not** be normalized or broken. This direct link should work: https://github.com/sporkmonger/addressable/pull/492/files#diff-ebf4cdeb47084eeefe86a7b9e669d9b89adb907a1c5c19c2928233fabb2c4f86

The result is :green_circle: green on this branch.
With `main` branch the result is :red_circle: :
```
  1) Addressable::Template Partial expand with symbols partial expand with unicode values preserves other encodings when input is not unicode
     Failure/Error: unpacked = input.unpack("U*")
     
     ArgumentError:
       malformed UTF-8 character
     # ./lib/addressable/idna/pure.rb:119:in `unpack'
     # ./lib/addressable/idna/pure.rb:119:in `unicode_normalize_kc'
     # ./lib/addressable/template.rb:948:in `normalize_value'
     # ./lib/addressable/template.rb:809:in `block in transform_capture'
     # ./lib/addressable/template.rb:791:in `each'
     # ./lib/addressable/template.rb:791:in `inject'
     # ./lib/addressable/template.rb:791:in `transform_capture'
     # ./lib/addressable/template.rb:750:in `block in transform_partial_capture'
     # ./lib/addressable/template.rb:747:in `each'
     # ./lib/addressable/template.rb:747:in `inject'
     # ./lib/addressable/template.rb:747:in `transform_partial_capture'
     # ./lib/addressable/template.rb:529:in `block in partial_expand'
     # ./lib/addressable/template.rb:528:in `gsub!'
     # ./lib/addressable/template.rb:528:in `partial_expand'
     # ./spec/addressable/template_spec.rb:1017:in `block (4 levels) in <top (required)>'
```

Probably not a big use-case otherwise people would have complained more I suppose? but I thought it was worth "improving" that.

The only case which could be "degraded" by this new code as @dentarg noted is if people pass in actual unicode string wrongly declared as another encoding. In such case they won't be normalized, whereas in the previous code they would have been. I don't think that's big deal as it was kind of a side-effect initially, and restoring this "behavior" will mean breaking compatibility with some other encodings, and for people who are correctly passing their strings encoding on top of it.

@dentarg @sporkmonger does this makes sense?

Comment (sporkmonger) [reply lib/addressable/template.rb:912]: Thanks for being thorough and checking the new test against `main`.

Comment (sporkmonger) [reply lib/addressable/template.rb:912]: So I think this is the key part of the template spec related to expansion methods:
> It is only during the process of template expansion that a string of characters in a URI Template is REQUIRED to be processed as a sequence of Unicode code points.
I read that as probably disallowing the behavior specified by this test unfortunately?

Comment (jarthod) [reply lib/addressable/template.rb:915]: > Thanks for being thorough and checking the new test against main.

:+1:

> I read that as probably disallowing the behavior described by this test description?

Good point, as I quickly scanned the spec (which is new to me) I noticed they mentionned some flexibility ("it should be understood that templates occur in practice as a sequence of characters in whatever form or encoding is suitable for the context in which they occur") which is what I expected considering the world of URI so I stopped there. And I missed the part you're showing right after that which says that for **expansion**, unicode is required.

In that case it makes things simpler, so I changed the code to accept any string, force encode it as UTF-8 and normalize it (so similar to `main` behavior). I've updated the specs to match this: https://github.com/sporkmonger/addressable/pull/492/files#diff-ebf4cdeb47084eeefe86a7b9e669d9b89adb907a1c5c19c2928233fabb2c4f86
They now test explicitely that:
1. unicode is normalized even if passed in as the wrong encoding
2. invalid UTF-8 chars are rejected (with an exception) like before.

Another approach could be to try to `encode` the incoming string to UTF-8 instead of just `force_encoding`, meaning a valid ISO string for example, would be converted to a different but valid UTF-8 string. But I think it's better to leave this behavior to the client as it could be too surprising. Also it could break some existing code usage were people already pass valid unicode string in the wrong encoding.

Comment (jarthod) [reply lib/addressable/template.rb:915]: @sporkmonger gentle ping :ping_pong: just to see if you agree, I'll start working on the other MR once this one is merged.

**Review by joseportillam — APPROVED** [review]

File: lib/addressable/template.rb
Comment (jarthod) [reply lib/addressable/template.rb:915]: @dentarg shall we keep waiting for Bob approval on this one? I don't want to be too pushy.

Comment (dentarg) [reply lib/addressable/template.rb:915]: @jarthod I can give him a nudge via email

Big disclaimer: I have not read the different RFCs as you two have, but I think your last comment makes sense.

Comment (jarthod) [reply lib/addressable/template.rb:915]: Ok thanks, there's no rush for me so better be safe and wait for @sporkmonger approval then, these RFCs are tricky so 2.5 pairs of eyes is always better :)

**Review by sporkmonger — APPROVED** [review]
Looks good to me!

**Comment by sporkmonger** [comment]
Sorry about being slow to respond! I was out on vacation last week and the laptop stayed home.

**Comment by jarthod** [comment]
@sporkmonger no problem, thank you!

File: lib/addressable/idna/native.rb
Comment (dentarg) [inline lib/addressable/idna/native.rb:34]: @jarthod Maybe we were a bit aggressive removing the method `Addressable::IDNA.unicode_normalize_kc` in a patch version. There might be code in the wild using these methods. Just happened to see [some code using `Addressable::IDNA.to_ascii`](https://github.com/discourse/discourse/pull/17840/files#diff-f98dab56f29862edb3b9003dde7224a286c23262bef7328698c4864c4ade1c6fR157).

Do you think it is possible to add it back? (In both native and pure versions, we don't know what people are using)

Comment (dentarg) [reply lib/addressable/idna/native.rb:34]: Would be cool to mark it as deprecated (though I don't think this project as an established way of doing that, some new code would be needed)

Comment (jarthod) [reply lib/addressable/idna/native.rb:34]: Why not, as you prefer. We can add it back easily, yes, with some caveats:
1. For `Pure` I don't think we should restore the old ruby implementation but use the new one (ruby's implementation) so it won't be 100% compatible with before but at least the method will exist.
2. Same thing for `Native`, if we restore the `IDN::Stringprep` one (meaning 100% identical to before the removal) the result will now be different than what addressable uses (ruby normalize). Also there's [no equivalent in `libidn2` AFAICS](https://libidn.gitlab.io/libidn2/reference/api-index-full.html). So we should also forward to ruby normalize I suppose.

---

For the deprecation, I didn't see any existing way either in the project. From a quick look I found this std ruby module: https://ruby-doc.org/stdlib-3.1.0/libdoc/rubygems/rdoc/Gem/Deprecate.html
Which seems to be [recommended](https://stackoverflow.com/a/23554720/304434) to do that (never used it). It prints this standard message we saw many times, would look like this (manually replacing the interpolations):
```
NOTE: Addressable::IDNA.unicode_normalize_kc is deprecated; use String#unicode_normalize(:nfkc) instead. It will be removed on or after 2023-04.
Addressable::IDNA.unicode_normalize_kc called from xxxxxxxx:42
```
The date is not optional unfortunately, but the message says "or after" so it's not a big deal we can leave "2023-04" so people know it's deprecated and will stop working at some point. This message allows to provide the replacement code which is good. Also it support the _standard_ `Gem::Deprecate.skip` option.

Alternatively we can [use `Kernel.warn` directly](https://stackoverflow.com/a/294114/304434), which even has a ["category" option which can take the value `:deprecated`](https://ruby-doc.org/3.2.1/Kernel.html#method-i-warn), though it's not clear what it changes and even Gem::Deprecate doesn't use it.

Comment (dentarg) [reply lib/addressable/idna/native.rb:34]: Thanks yet again for an awesome comment @jarthod :) I agree with you that we should not restore the old behaviours, just bring back the methods so code doesn't break. I'm fine with either options for deprecating.

Comment (jarthod) [reply lib/addressable/idna/native.rb:34]: → #504

**Pull request merged**
